### PR TITLE
Asyncs thread

### PR DESCRIPTION
**Pull request by lixinqi**
完整实现 flow.asyncs.Thread/flow.asyncs.thread

### CODE REVIEW

File: python/oneflow/test/modules/test_asyncs_thread.py [hunk 1]
```diff
@@ -27,54 +27,38 @@
...
             test_case.assertEqual(tensor[int(tensor.shape[0] / 2)], 1 + i)
 
-    @unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test cpu cases")
-    def test_decorator(test_case):
-        thread_ids = [i % 4 for i in range(200)]
-        tensors = []
-        dim = 0
-
-        def Func():
-            placement = flow.placement("cuda", [0, 1])
```
Comment (lixinqi) [inline python/oneflow/test/modules/test_asyncs_thread.py:67]: 这个测试是每个rank开7个线程。每个线程不断用0,1,2,3 comm_id与其他rank的对应线程做通信操作。此处设备类为cuda会导致bug，而如果是cpu就不会正常执行下来。看起来所有的nccl操作都必须顺序化才行。

File: python/oneflow/test/modules/test_asyncs_thread.py [hunk 2]
```diff
@@ -27,54 +27,38 @@
...
+        threads = [flow.asyncs.Thread() for i in range(7)]
+        iter_and_thread_and_comm_ids = [(i, threads[i % 7], i % 4) for i in range(200)]
         tensors = []
         dim = 0
-        for i, thread in enumerate(thread_ids):
+        for i, thread, comm_id in iter_and_thread_and_comm_ids:
             dim += 1
-            with flow.asyncs.thread(thread):
+            with flow.asyncs.thread(thread, comm_id):
                 placement = flow.placement("cuda", [0, 1])
```
Comment (lixinqi) [inline python/oneflow/test/modules/test_asyncs_thread.py:55]: 这个测试是每个rank开7个线程。每个线程不断用0,1,2,3 comm_id与其他rank的对应线程做通信操作。此处设备类为cuda会导致bug，而如果是cpu就不会正常执行下来。看起来所有的nccl操作都必须顺序化才行。

**Comment by lixinqi** [comment]
多个comm_id会导致bug，输出如下栈：
```
F20220901 22:34:16.024556 2482634 rpc_client.cpp:40] Check failed: stub->CallMethod<ctrl_method>(&client_ctx, request_, &response_).error_code() == grpc::StatusCode::OK (14 vs. 0)
*** Check failure stack trace: ***
F20220901 22:34:16.024561 2482629 rpc_client.cpp:40] Check failed: stub->CallMethod<ctrl_method>(&client_ctx, request_, &response_).error_code() == grpc::StatusCode::OK (14 vs. 0) F20220901 22:34:16.024570 2482633 rpc_client.cpp:40] Check failed: stub->CallMethod<ctrl_method>(&client_ctx, request_, &response_).error_code() == grpc::StatusCode::OK (14 vs. 0) F20220901 22:34:16.024703 2482631 rpc_client.cpp:40] Check failed: stub->CallMethod<ctrl_method>(&client_ctx, request_, &response_).error_code() == grpc::StatusCode::OK (14 vs. 0)
*** Check failure stack trace: ***
F20220901 22:34:16.024561 2482629 rpc_client.cpp:40] Check failed: stub->CallMethod<ctrl_method>(&client_ctx, request_, &response_).error_code() == grpc::StatusCode::OK (14 vs. 0) F20220901 22:34:16.024570 2482633 rpc_client.cpp:40] Check failed: stub->CallMethod<ctrl_method>(&client_ctx, request_, &response_).error_code() == grpc::StatusCode::OK (14 vs. 0) F20220901 22:34:16.024703 2482631 rpc_client.cpp:40] Check failed: stub->CallMethod<ctrl_method>(&client_ctx, request_, &response_).error_code() == grpc::StatusCode::OK (14 vs. 0)
*** Check failure stack trace: ***
F20220901 22:34:16.024561 2482629 rpc_client.cpp:40] Check failed: stub->CallMethod<ctrl_method>(&client_ctx, request_, &response_).error_code() == grpc::StatusCode::OK (14 vs. 0) F20220901 22:34:16.024570 2482633 rpc_client.cpp:40] Check failed: stub->CallMethod<ctrl_method>(&client_ctx, request_, &response_).error_code() == grpc::StatusCode::OK (14 vs. 0) F20220901 22:34:16.024703 2482631 rpc_client.cpp:40] Check failed: stub->CallMethod<ctrl_method>(&client_ctx, request_, &response_).error_code() == grpc::StatusCode::OK (14 vs. 0)
*** Check failure stack trace: ***
    @     0x7f39475be355  google::LogMessage::Flush()
    @     0x7f39475be355  google::LogMessage::Flush()
    @     0x7f39475be355  google::LogMessage::Flush()
    @     0x7f39475be355  google::LogMessage::Flush()
    @     0x7f39475c168c  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f39475c168c  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f393cb7e6f7  oneflow::(anonymous namespace)::ClientCall<>::operator()()
    @     0x7f393cb7e6f7  oneflow::(anonymous namespace)::ClientCall<>::operator()()
    @     0x7f39475c168c  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f39475c168c  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f393cb7e6f7  oneflow::(anonymous namespace)::ClientCall<>::operator()()
    @     0x7f393cb7e6f7  oneflow::(anonymous namespace)::ClientCall<>::operator()()
    @     0x7f393cb7e5a5  oneflow::RpcClient::PullKV()
    @     0x7f393cb7e5a5  oneflow::RpcClient::PullKV()
    @     0x7f393cb7e5a5  oneflow::RpcClient::PullKV()
    @     0x7f393cb7e5a5  oneflow::RpcClient::PullKV()
    @     0x7f393cb6fb80  oneflow::GrpcCtrlClient::PullKV()
    @     0x7f393cb6fb80  oneflow::GrpcCtrlClient::PullKV()
    @     0x7f393cb6fb80  oneflow::GrpcCtrlClient::PullKV()
    @     0x7f39403381bd  oneflow::(anonymous namespace)::CreateNcclComm()
    @     0x7f393cb6fb80  oneflow::GrpcCtrlClient::PullKV()
    @     0x7f39403381bd  oneflow::(anonymous namespace)::CreateNcclComm()
    @     0x7f39403381bd  oneflow::(anonymous namespace)::CreateNcclComm()
    @     0x7f39403381bd  oneflow::(anonymous namespace)::CreateNcclComm()
    @     0x7f3940337c8b  oneflow::EagerNcclCommMgr::GetCommForDevice()
    @     0x7f3940337c8b  oneflow::EagerNcclCommMgr::GetCommForDevice()
    @     0x7f3940337c8b  oneflow::EagerNcclCommMgr::GetCommForDevice()
    @     0x7f3940337c8b  oneflow::EagerNcclCommMgr::GetCommForDevice()
    @     0x7f39416335e1  oneflow::ccl::CudaCommunicationContext::Init()
    @     0x7f39416335e1  oneflow::ccl::CudaCommunicationContext::Init()
    @     0x7f39416335e1  oneflow::ccl::CudaCommunicationContext::Init()
    @     0x7f39416335e1  oneflow::ccl::CudaCommunicationContext::Init()
    @     0x7f3941a7915d  oneflow::ccl::NewCommunicationContext()
    @     0x7f3941a7915d  oneflow::ccl::NewCommunicationContext()
    @     0x7f3941a7915d  oneflow::ccl::NewCommunicationContext()
    @     0x7f3941a781e5  oneflow::(anonymous namespace)::InitEagerCclOpKernelCache()
    @     0x7f3941a7915d  oneflow::ccl::NewCommunicationContext()
    @     0x7f3941a781e5  oneflow::(anonymous namespace)::InitEagerCclOpKernelCache()
    @     0x7f3941a781e5  oneflow::(anonymous namespace)::InitEagerCclOpKernelCache()
    @     0x7f3941a781e5  oneflow::(anonymous namespace)::InitEagerCclOpKernelCache()
    @     0x7f394307448a  oneflow::one::StatefulOpKernel::TryInitOpKernelStateAndCache()
    @     0x7f394307448a  oneflow::one::StatefulOpKernel::TryInitOpKernelStateAndCache()
    @     0x7f394307448a  oneflow::one::StatefulOpKernel::TryInitOpKernelStateAndCache()
    @     0x7f394307448a  oneflow::one::StatefulOpKernel::TryInitOpKernelStateAndCache()
    @     0x7f394119e5a0  oneflow::vm::OpCallInstructionUtil::TryInitOpKernelStateAndCache()
    @     0x7f394119e5a0  oneflow::vm::OpCallInstructionUtil::TryInitOpKernelStateAndCache()
    @     0x7f394119e5a0  oneflow::vm::OpCallInstructionUtil::TryInitOpKernelStateAndCache()
    @     0x7f394119e5a0  oneflow::vm::OpCallInstructionUtil::TryInitOpKernelStateAndCache()
    @     0x7f394119c554  oneflow::vm::OpCallInstructionPolicy::Compute()
    @     0x7f394119c554  oneflow::vm::OpCallInstructionPolicy::Compute()
    @     0x7f394119c554  oneflow::vm::OpCallInstructionPolicy::Compute()
    @     0x7f394119c554  oneflow::vm::OpCallInstructionPolicy::Compute()
    @     0x7f39411998ef  oneflow::vm::InstructionPolicy::ComputeIf()
    @     0x7f39411998ef  oneflow::vm::InstructionPolicy::ComputeIf()
    @     0x7f39411998ef  oneflow::vm::InstructionPolicy::ComputeIf()
    @     0x7f39411998ef  oneflow::vm::InstructionPolicy::ComputeIf()
    @     0x7f394119882f  oneflow::vm::EpStreamPolicyBase::Run()
    @     0x7f394119882f  oneflow::vm::EpStreamPolicyBase::Run()
    @     0x7f394119882f  oneflow::vm::EpStreamPolicyBase::Run()
    @     0x7f394119882f  oneflow::vm::EpStreamPolicyBase::Run()
    @     0x7f39411a09b7  oneflow::vm::StreamPolicy::RunIf()
    @     0x7f39411a09b7  oneflow::vm::StreamPolicy::RunIf()
    @     0x7f39411a09b7  oneflow::vm::StreamPolicy::RunIf()
    @     0x7f39411a09b7  oneflow::vm::StreamPolicy::RunIf()
    @     0x7f39411a2150  oneflow::vm::ThreadCtx::TryReceiveAndRun()
    @     0x7f39411a2150  oneflow::vm::ThreadCtx::TryReceiveAndRun()
    @     0x7f39411a2150  oneflow::vm::ThreadCtx::TryReceiveAndRun()
    @     0x7f39411af288  oneflow::(anonymous namespace)::WorkerLoop()
    @     0x7f39411af288  oneflow::(anonymous namespace)::WorkerLoop()
    @     0x7f39411a2150  oneflow::vm::ThreadCtx::TryReceiveAndRun()
    @     0x7f39411af288  oneflow::(anonymous namespace)::WorkerLoop()
    @     0x7f39411b2424  std::thread::_State_impl<>::_M_run()
    @     0x7f390b4abde4  (unknown)
    @     0x7f39411b2424  std::thread::_State_impl<>::_M_run()
    @     0x7f39411af288  oneflow::(anonymous namespace)::WorkerLoop()
    @     0x7f39411b2424  std::thread::_State_impl<>::_M_run()
    @     0x7f390b4abde4  (unknown)
    @     0x7f3baeeba609  start_thread
    @     0x7f390b4abde4  (unknown)
    @     0x7f3baeeba609  start_thread
    @     0x7f3baeddf163  clone
Aborted

```
这是四个线程同时打印出来的栈信息

**Comment by lixinqi** [comment]
大体的原因是EagerNcclCommMgr::GetCommForDevice里边每个线程在调用GrpcClient的PUSH/POP所用的key是一摸一样的，导致了问题。

**Comment by lixinqi** [comment]
> 大体的原因是EagerNcclCommMgr::GetCommForDevice里边每个线程在调用GrpcClient的PUSH/POP所用的key是一摸一样的，导致了问题。

如果不改EagerNcclCommMgr，必须得让所有的nccl操作强制顺序执行才能不出bug。

**Comment by lixinqi** [comment]
如果重构EagerNcclCommMgr，把comm_id编码到CtrlClient pull/push 的key里，就会解决 https://github.com/Oneflow-Inc/oneflow/pull/9039#issuecomment-1234986458 这里所提的问题。但是仍然会出现新的问题：
```
oneflow-22:2804723:2804900 [0] enqueue.cc:630 NCCL WARN Error : mixing different streams within a group call is not supported.
oneflow-22:2804723:2804900 [0] NCCL INFO enqueue.cc:1365 -> 5
F20220902 10:16:13.060290 2804900 cuda_all_gather.cpp:40] Check failed: ncclAllGather(in, out, elem_cnt, nccl_datatype_, cuda_communication_ctx->nccl_comm(), stream->As<ep::CudaStream>()->cuda_stream()) : invalid usage (5). To see more detail, please run OneFlow with system variable NCCL_DEBUG=INFO
```

**Comment by lixinqi** [comment]
https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/communicators.html
### Using multiple NCCL communicators concurrently
Using multiple NCCL communicators requires careful synchronization, or can lead to deadlocks.

NCCL kernels are blocking (waiting for data to arrive), and any CUDA operation can cause a device synchronization, meaning it will wait for all NCCL kernels to complete. This can quickly lead to deadlocks since NCCL operations perform CUDA calls themselves.

Operations on different communicators should therefore be used at different epochs with a locking mechanism, and applications should ensure operations are submitted in the same order across ranks.

Launching multiple communication operations (on different streams) might work provided they can fit within the GPU, but could break at any time if NCCL were to use more CUDA blocks per operation, or if some calls used inside NCCL collectives were to perform a device synchronization (e.g. allocate some CUDA memory dynamically).

nccl看起来不希望多个线程同时使用不同的communicator。我们要不要将flow.asyncs.thread参数comm_id改为host_comm_id？

**Comment by lixinqi** [comment]
去掉了comm_id，所有的传输类op都顺序化。

File: oneflow/api/python/framework/thread.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+namespace py = pybind11;
+
+namespace oneflow {
+
+namespace {
+
+class UsingThreadUidSet final {
+ public:
+  explicit UsingThreadUidSet() {
```
Comment (strint) [inline oneflow/api/python/framework/thread.cpp:29]: 无参情况下，explicit 是不是没有作用，因为没有类型转换的可能？

File: oneflow/api/python/framework/thread.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+#define ONEFLOW_API_PYTHON_FRAMEWORK_THREAD_H_
+
+#include "oneflow/core/framework/stream.h"
+
+namespace oneflow {
+
+class AsyncThread final {
+ public:
+  AsyncThread(const AsyncThread&) = delete;
+  AsyncThread(AsyncThread&&) = delete;
```
Comment (strint) [inline oneflow/api/python/framework/thread.h:26]: 现在不提倡用 DISABLE_COPY_AND_MOVE 那个宏了么

**Review by strint — APPROVED** [review]

File: oneflow/api/python/framework/thread.h (hunk 4 above)
Comment (lixinqi) [reply oneflow/api/python/framework/thread.h]: 已处理：https://github.com/Oneflow-Inc/oneflow/pull/9039/commits/dc2c4fb8e07e720596e4d4296c640da4256cd5e4

我个人习惯不引用。因为也没多少代码，反而增加了一个include。

File: oneflow/api/python/framework/thread.cpp (hunk 3 above)
Comment (lixinqi) [reply oneflow/api/python/framework/thread.cpp]: 已处理：https://github.com/Oneflow-Inc/oneflow/pull/9039/commits/dc2c4fb8e07e720596e4d4296c640da4256cd5e4

File: oneflow/core/common/optional.h [hunk 5]
```diff
@@ -455,4 +457,20 @@ class Optional final : private internal::OptionalBase<T> {
 
 }  // namespace oneflow
 
+namespace std {
+
+template<typename T>
+struct hash<oneflow::Optional<T>> final {
```
[the remark this answers is not in the document]
Comment (lixinqi) [reply oneflow/core/common/optional.h]: 已去掉：https://github.com/Oneflow-Inc/oneflow/pull/9039/commits/5e073364ced4aff654cf212f58a26262ee55bdc8

**Review by hjchen2 — APPROVED** [review]

**Pull request merged**